keps/sig-cluster-lifecycle/kubeadm/20190925-component-configs.md

```diff
@@ -0,0 +1,223 @@
+There is support for users to supply their own component configs along with the kubeadm specific config objects upon kubeadm init. Kubeadm validates, defaults and stores these config on the cluster in the form of config maps. It also supplies these configs in one way or another to the target component itself.  
+There is also a desire and code has been historically structured in a way, to allow kubeadm to convert component config formats upon upgrade, should the need for that arise.
+
+Hence, the following problems have been observed:
+- Internal types have been used to allow for defaulting, validating and migrating component configs. This cannot continue if kubeadm is to move out of the kubernetes/kubernetes repository.
+- Defaulting the configs, that are generated by kubeadm or supplied by users, populates additional fields with default values. This makes it difficult to determine in retrospect, what fields were set on purpose with their default values, and what fields were simply automatically filled in by the defaulting code.
+- Providing support for validation by vendoring component specific code can lead to false-positive and missed errors. This can be caused by the deviation between the imported code by kubeadm and the actual component.
+- Migrating between config versions also requires code to be vendored in or forked from the components. This can also lead to difficult to diagnose errors and changed behaviour in unexpected ways.
+- Not distinguishing between kubeadm generated and user supplied component configs, necessitates the requirement for config migration. Kubeadm generated configs that are not modified by users directly in the config map, can be fully regenerated upon config format change. In all other cases kubeadm must convert the config to preserve user changes.
+
+These observations along with the emergence of kubeadm phases, kustomize support and work around addon operators allows us to have the following future goals.
+
+### Goals
+
+- Stop vendoring internal types and validation, defaulting and conversion code.
+- Stop defaulting component configs.
+- Outsource validation to the component binaries themselves.
+- Distinguish between user supplied and kubeadm generated and unmodified component configs.
+- Eradicate the need to convert configs
+- Use more community wide solutions, like Kustomize, kubeadm phases and addon operators
+
+### Non-Goals
+
+- Change the handling of kubeadm’s own config types in any way.
+- Cover in any way the migration of command line flags to component configs.
+- Define strategies for config map naming and backup.
+
+## Proposal
+
+This is where we get down to the nitty gritty of what the proposal actually is.
```

Comment:
this sentence can be omitted?

Reply:
My bad, copy-paste error from the template.